### PR TITLE
Allow requesting output ids without pagination

### LINKED ISSUE #1541: Allow requesting output ids from indexer without automatic pagination
## Description

Allow requesting output ids from indexer without pagination

## Motivation

Currently the requests to the indexer use `get_output_ids_with_pagination()` internally, which always requests all output ids that are available
This could lead to errors because of too many requests or also might be unnecessary if not all outputs are wanted (maybe no query parameters are wanted)

## Requirements

1. Add functions to request output ids from indexer without pagination
2. Accept a cursor so the certain pages can be requested manually
3. Return `OutputIdsResponse` so the cursor is also available

## Are you planning to do it yourself in a pull request?

No.

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Allow requesting output ids without pagination, because it wasn't possible before, which could lead to useless requests or even timeouts

## Links to any relevant issues

Fixes https://github.com/iotaledger/iota.rs/issues/1541

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Running examples/tests

## Change checklist

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by thibault-martinez** [comment]
The requirements on the issue say `Accept a cursor so the certain pages can be requested manually`, shouldn't the methods take a `Option<Cursor>` instead of a boolean?

**Comment by Thoralf-M** [comment]
The cursor can already be provided as query parameter `QueryParameter::Cursor(String)`

**Comment by thibault-martinez** [comment]
> The cursor can already be provided as query parameter `QueryParameter::Cursor(String)`

Shouldn't it be based on the presence/absence of it instead of an additional bool then?

**Comment by Thoralf-M** [comment]
> > The cursor can already be provided as query parameter `QueryParameter::Cursor(String)`
> 
> Shouldn't it be based on the presence/absence of it instead of an additional bool then?

Would be also an option, but what if you have the first 10 pages, the cursor for the 11. and want all remaining? If you provide the cursor and we only return that page, then you would need to do the loop until you got all on your side

File: client/src/api/high_level.rs [hunk 1]
```diff
@@ -266,7 +266,7 @@ impl Client {
                 ])
                 .await?;
 
-            output_responses.extend(self.get_outputs(basic_output_ids).await?);
+            output_responses.extend(self.get_outputs(basic_output_ids.items).await?);
```
Comment (thibault-martinez) [inline client/src/api/high_level.rs]: You renamed them all except this one

File: client/CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security -->
 
+## 2.0.1-rc.8 - 2023-XX-XX
+
+### Changed
+
+- `OutputIdsResponse::items` from `Vec<String>` to `Vec<OutputId>`;
```
Comment (thibault-martinez) [inline client/CHANGELOG.md]: Should be in the types changelog

File: client/src/node_api/indexer/routes.rs [hunk 3]
```diff
@@ -2,86 +2,50 @@
 // SPDX-License-Identifier: Apache-2.0
 
 //! IOTA node indexer routes
-use iota_types::block::output::{AliasId, FoundryId, NftId, OutputId};
-
-use crate::{node_api::indexer::query_parameters::QueryParameter, Client, Error, Result};
+use iota_types::{
```
Comment (thibault-martinez) [inline client/src/node_api/indexer/routes.rs:6]:
```suggestion

use iota_types::{
```

File: client/src/node_api/indexer/routes.rs [hunk 4]
```diff
@@ -90,7 +55,7 @@ impl Client {
         let route = format!("api/indexer/v1/outputs/alias/{alias_id}");
 
         Ok(*(self
-            .get_output_ids_with_pagination(&route, Vec::new(), true, false)
+            .get_output_ids(&route, QueryParameters::new(Vec::new()), true, false)
```
Comment (thibault-martinez) [inline client/src/node_api/indexer/routes.rs]: Maybe a `QueryParameters::empty()` method would be handy

File: client/src/node_api/indexer/mod.rs [hunk 5]
```diff
@@ -6,27 +6,32 @@
...
+use iota_types::api::plugins::indexer::OutputIdsResponse;
 
 pub(crate) use self::query_parameters::{QueryParameter, QueryParameters};
 use crate::{Client, Result};
 
 impl Client {
     /// Get all output ids for a provided URL route and query parameters.
-    pub async fn get_output_ids_with_pagination(
+    /// If a `QueryParameter::Cursor(_)` is provided, only a single page will be queried.
+    pub async fn get_output_ids(
```
Comment (thibault-martinez) [inline client/src/node_api/indexer/mod.rs:17]: The rename is not in the changelog

File: client/examples/node_api_indexer/07_get_random_basic_outputs.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        std::env::var("NODE_URL").unwrap()
+    });
+
+    // Create a client with that node.
+    let client = Client::builder()
+        // The node needs to have the indexer plugin enabled.
+        .with_node(&node_url)?
+        .finish()?;
+
+    // Get a single page with random output IDs by providing only `QueryParameter::Cursor(_)`.
```
Comment (thibault-martinez) [inline client/examples/node_api_indexer/07_get_random_basic_outputs.rs:24]: What page would that be ?

File: client/CHANGELOG.md [hunk 7]
```diff
@@ -19,6 +19,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security -->
 
+## 2.0.1-rc.8 - 2023-XX-XX
+### Added
```
Comment (thibault-martinez) [inline client/CHANGELOG.md:24]:
```suggestion

### Added
```

**Review by thibault-martinez — APPROVED** [review]

File: client/examples/node_api_indexer/07_get_random_basic_outputs.rs (hunk 6 above)
Comment (Thoralf-M) [reply client/examples/node_api_indexer/07_get_random_basic_outputs.rs:24]: The first one

Comment (thibault-martinez) [reply client/examples/node_api_indexer/07_get_random_basic_outputs.rs:24]: Do we (client) convert empty cursor to first page or is that an indexer thing ?

Comment (Thoralf-M) [reply client/examples/node_api_indexer/07_get_random_basic_outputs.rs:24]: Indexer uses the first page by default, unless you provided a valid cursor for another page

**Pull request merged**